README
Complete documentation for CL-INTERPOL can be found in the 'doc'
directory.

CL-INTERPOL also supports Nikodemus Siivola's HYPERDOC, see
<http://common-lisp.net/project/hyperdoc/> and
<http://www.cliki.net/hyperdoc>.

1. Installation

1.1. Probably the easiest way is

       (load "/path/to/cl-interpol/load.lisp")

     This should compile and load CL-INTERPOL on most Common Lisp
     implementations.

1.2. With MK:DEFSYSTEM you can make a symbolic link from
     'cl-interpol.system' and 'cl-interpol-test.system' to your central registry
     (which by default is in '/usr/local/lisp/Registry/') and then issue
     the command

       (mk:compile-system "cl-interpol")

     Note that this relies on TRUENAME returning the original file a
     symbolic link is pointing to. This will only work with AllegroCL
     6.2 if you've applied all patches with (SYS:UPDATE-ALLEGRO).

1.3. You can also use ASDF instead of MK:DEFSYSTEM in a similar way
     (use the .asd files instead of the .system files).

In order to actually use CL-INTERPOL you have to enable its reader
syntax with CL-INTERPOL:ENABLE-INTERPOL-SYNTAX - see the docs.

2. Test

For a quick test load the file "test.lisp" after installing
CL-INTERPOL. (Note: SBCL will emit tons of compiler notes when loading
"test.lisp" but it will eventually, after a couple of minutes, pass
all tests - at least it does for me with 0.8.4.8.)


CHANGELOG
Version 0.2.7
2018-05-08
Added support for name-readtables.
Now interpol syntax can be activated using: (named-readtables:in-readtable :interpol-syntax).
This way integration with SLIME and SLY is possible.

Version 0.2.6
2016-08-26
Merge pull request #5 from agrostis/master (Hans Hübner)
Followup to 100efc6: added documentation and tests. (Boris Smilga)
Followup to 100efc6: ignore non-newline whitespaces only, as per CLHS 22.3.9.3. (Boris Smilga)
Merge pull request #4 from agrostis/master (Hans Hübner)
Backspace + newline as an escape sequence emulating tilde + newline in CL:FORMAT. (Boris Smilga)

Version 0.2.5
2015-11-22
Fix docstring as suggested by Stas (Hans Hübner)

Version 0.2.4
2015-11-22
Exnabled direct use of cl-interpol-reader (Mike Maul)

Version 0.2.3
2014-11-28
update support info (Hans Huebner)

Version 0.2.2
2013-10-04
inline format directive syntax and an extra argument to enable-interpol-syntax (Marco Baringer)

Version 0.2.1
2008-07-25
Fixed typo in HTML documentation

Version 0.2.0
2008-07-24
Base Unicode support on CL-UNICODE
Add new CL-PPCRE special characters for named registers and named properties
Re-architecture test suite

Version 0.1.2
2004-12-16
Added hyperdoc support
Added :CL-INTERPOL to *FEATURES*
Typo fixes in doc/index.html

Version 0.1.1
2003-12-21
Fixed an embarrassing bug where COLLECTOR was re-used in read.lisp (reported by Hans Hübner)
More tests, better failure reporting
Tried to increase readability of docs
Mentioned Debian and Gentoo in docs

Version 0.1.0
2003-10-22
Initial release


test.pl
#!/usr/bin/perl

### $Header: /usr/local/cvsrep/cl-interpol/test.pl,v 1.4 2003/10/19 15:55:25 edi Exp $

### Copyright (c) 2002-2003, Dr. Edmund Weitz. All rights reserved.

### Redistribution and use in source and binary forms, with or without
### modification, are permitted provided that the following conditions
### are met:

###   * Redistributions of source code must retain the above copyright
###     notice, this list of conditions and the following disclaimer.

###   * Redistributions in binary form must reproduce the above
###     copyright notice, this list of conditions and the following
###     disclaimer in the documentation and/or other materials
###     provided with the distribution.

### THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
### OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
### WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
### ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
### DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
### DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
### GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
### INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
### WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
### NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
### SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

my @chars = qw(\Q \L \U \E \l \u);

sub combine {
  my $delim = shift;
  my @result = ();
  foreach my $char (@chars) {
    foreach my $string (@_) {
      push @result, "$char$delim$string";
    }
  }
  @result;
}

sub quote {
  local $_ = shift;
  s/\\/\\\\/g;
  $_;
}

print <<'HEAD';
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL-INTERPOL; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/cl-interpol/test.pl,v 1.4 2003/10/19 15:55:25 edi Exp $

;;; Copyright (c) 2002-2003, Dr. Edmund Weitz. All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

;;; This file was automatically generated by a Perl script. It should
;;; not be loaded directly but by "load.lisp" instead.

(in-package #:cl-interpol)

(enable-interpol-syntax)

HEAD

foreach my $a (('Aa-', 'aA-')) {
  my $counter = 0;
  my @arr = @chars;
  while ($counter++ < 4) {
    foreach my $str (@arr) {
      print "(let ((a \"$a\"))\n";
      my $test = "\${a}$str\${a}";
      print "  (test #?\"$test\" \"" . (quote eval "\"$test\"") . "\"))\n";
    }
    @arr = combine '${a}', @arr;
  }
}

print "\n(disable-interpol-syntax)\n";


test/tests.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL-INTERPOL-TEST; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/cl-interpol/test/tests.lisp,v 1.4 2008/07/23 16:10:13 edi Exp $

;;; Copyright (c) 2008, Dr. Edmund Weitz. All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(in-package :cl-interpol-test)

(defvar *this-file* (load-time-value
                     (or #.*compile-file-pathname* *load-pathname*))
  "The location of this source file.  Needed to find the data files.")

(defmacro do-tests ((name &optional show-progress-p) &body body)
  "Helper macro which repeatedly executes BODY until the code in body
calls the function DONE.  It is assumed that each invocation of BODY
will be the execution of one test which returns NIL in case of success
and a list of strings describing errors otherwise.

The macro prints a simple progress indicator \(one dots for ten tests)
to *STANDARD-OUTPUT* unless SHOW-PROGRESS-P is NIL and returns a true
value iff all tests succeeded.  Errors in BODY are caught and reported
\(and counted as failures)."
  `(let ((successp t)
         (testcount 1))
     (block test-block
       (flet ((done ()
                (return-from test-block successp)))
         (format t "~&Test: ~A~%" ,name)
         (loop
          (when (and ,show-progress-p (zerop (mod testcount 10)))
            (format t ".")
            (when (zerop (mod testcount 100))
              (terpri))
            (force-output))
          (let ((errors
                 (handler-case
                     (progn ,@body)
                   (error (msg)
                     (list (format nil "~&got an unexpected error: ~A" msg))))))
            (setq successp (and successp (null errors)))
            (when errors
              (format t "~&~4@A:~{~&   ~A~}~%" testcount errors))
            (incf testcount)))))
     successp))

(defun simple-tests (&key (file-name
                           (make-pathname :name "simple"
                                          :type nil :version nil
                                          :defaults *this-file*))                          
                          (external-format '(:latin-1 :eol-style :lf))
                          verbose)
  "Loops through all the forms in the file FILE-NAME and executes each
of them using EVAL.  The CL-INTERPOL syntax is enabled when the forms
are read.  It is assumed that each FORM specifies a test which returns
a true value iff it succeeds.  Prints each test form to
*STANDARD-OUTPUT* if VERBOSE is true and shows a simple progress
indicator otherwise.  EXTERNAL-FORMAT is the FLEXI-STREAMS external
format which is used to read the file.  Returns a true value iff all
tests succeeded.

\(SETQ ...) forms are treated special in that they're just EVALuated
but not counted as tests.  The global special variables exported by
CL-INTERPOL \(and some from CL-UNICODE as well) are rebound during the
tests so that they can be safely set in the test files."
  (enable-interpol-syntax)
  (unwind-protect
      (with-open-file (binary-stream file-name :element-type 'flex:octet)
        (let* ((stream (flex:make-flexi-stream binary-stream :external-format external-format))
               (*package* (find-package :cl-interpol-test))
               (*list-delimiter* *list-delimiter*)
               (*outer-delimiters* *outer-delimiters*)
               (*inner-delimiters* *inner-delimiters*)
               (*optional-delimiters-p* *optional-delimiters-p*)
               (*scripts-to-try* *scripts-to-try*)
               (*try-abbreviations-p* *try-abbreviations-p*)
               (*try-hex-notation-p* *try-hex-notation-p*)
               (*try-lisp-names-p* *try-lisp-names-p*)
               (*try-unicode1-names-p* *try-unicode1-names-p*))
          (do-tests ((format nil "Simple tests from file ~S" (file-namestring file-name))
                     (not verbose))
            (let* ((form (or (read stream nil) (done)))
                   (setqp (eq (first form) 'setq)))
              (when (and verbose (not setqp))
                (format t "~&~S" form))
              (cond (setqp (eval form) nil)
                    ((eval form) nil)
                    (t (list (format nil "~S returned NIL" form))))))))
    (disable-interpol-syntax)))

(defun run-all-tests (&key verbose)
  "Runs all tests for CL-INTERPOL and returns a true value iff all
tests succeeded.  VERBOSE is interpreted by the individual test suites
above."
  (let ((successp t))
    (macrolet ((run-test-suite (&body body)
                 `(unless (progn ,@body)
                    (setq successp nil))))
      ;; run the automatically generated Perl tests
      (run-test-suite (simple-tests :file-name (make-pathname :name "perltests"
                                                              :type nil :version nil
                                                              :defaults *this-file*)
                                    :verbose verbose))
      (run-test-suite (simple-tests :verbose verbose)))
    (format t "~2&~:[Some tests failed~;All tests passed~]." successp)
    successp))


util.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL-INTERPOL; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/cl-interpol/util.lisp,v 1.7 2003/10/22 09:20:30 edi Exp $

;;; Copyright (c) 2003, Dr. Edmund Weitz. All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(in-package #:cl-interpol)

(define-condition simple-reader-error (simple-condition reader-error)
  ()
  (:documentation "A reader error which can be signalled by ERROR."))

(defmacro signal-reader-error (format-control &rest format-arguments)
  "Like ERROR but signals a SIMPLE-READER-ERROR for the stream
*STREAM*."
  `(error 'simple-reader-error
          :stream *stream*
          :format-control ,format-control
          :format-arguments (list ,@format-arguments)))

(defun string-list-to-string (string-list)
  "Concatenates a list of strings to one string."
  ;; this function originally provided by JP Massar for CL-PPCRE; note
  ;; that we can't use APPLY with CONCATENATE here because of
  ;; CALL-ARGUMENTS-LIMIT
  (let ((total-size 0))
    (dolist (string string-list)
      (incf total-size (length string)))
    (let ((result-string (make-array total-size :element-type 'character))
          (curr-pos 0))
      (dolist (string string-list)
        (replace result-string string :start1 curr-pos)
        (incf curr-pos (length string)))
      result-string)))

(defun quote-meta-chars (string)
  "Quote, i.e. prefix with #\\\\, all non-word characters in STRING."
  (with-output-to-string (s)
    (loop for char across string
          if (or (char<= #\a char #\z)
                 (char<= #\A char #\Z)
                 (char<= #\0 char #\9)
                 (char= #\_ char)) do
          (write-char char s)
          else do
          (write-char #\\ s)
          (write-char char s))))

(defun get-end-delimiter (start-delimiter delimiters &key errorp)
  "Find the closing delimiter corresponding to the opening delimiter
START-DELIMITER in a list DELIMITERS which is formatted like
*OUTER-DELIMITERS*. If ERRORP is true, signal an error if none was
found, otherwise return NIL."
  (loop for element in delimiters
        if (eql start-delimiter element)
        do (return-from get-end-delimiter start-delimiter)
        else if (and (consp element)
                     (char= start-delimiter (car element)))
        do (return-from get-end-delimiter (cdr element)))
  (when errorp
    (signal-reader-error "~S not allowed as a delimiter here" start-delimiter)))

(declaim (inline make-collector))
(defun make-collector ()
  "Create an empty string which can be extended by
VECTOR-PUSH-EXTEND."
  (make-array 0
              :element-type 'character
              :fill-pointer t
              :adjustable t))

(declaim (inline make-char-from-code))
(defun make-char-from-code (number)
  "Create character from char-code NUMBER. NUMBER can be NIL which is
interpreted as 0."
  ;; Only look at rightmost eight bits in compliance with Perl
  (let ((code (logand #o377 (or number 0))))
    (or (and (< code char-code-limit)
             (code-char code))
        (signal-reader-error "No character for char-code #x~X"
                             number))))

(declaim (inline lower-case-p*))
(defun lower-case-p* (char)
  "Whether CHAR is a character which has case and is lowercase."
  (or (not (both-case-p char))
      (lower-case-p char)))

(defmacro read-char* ()
  "Convenience macro because we always read from the same string with
the same arguments."
  `(read-char *stream* t nil t))

(defmacro peek-char* ()
  "Convenience macro because we always peek at the same string with
the same arguments."
  `(peek-char nil *stream* t nil t))

(declaim (inline copy-readtable*))
(defun copy-readtable* ()
  "Returns a copy of the readtable which was current when
INTERPOL-READER was invoked. Memoizes its result."
  (or *readtable-copy*
      (setq *readtable-copy* (copy-readtable))))

(declaim (inline nsubvec))
(defun nsubvec (sequence start &optional (end (length sequence)))
  "Return a subvector by pointing to location in original vector."
  (make-array (- end start)
              :element-type (array-element-type sequence)
              :displaced-to sequence
              :displaced-index-offset start))


packages.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL-USER; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/cl-interpol/packages.lisp,v 1.11 2008/07/23 15:35:07 edi Exp $

;;; Copyright (c) 2003-2008, Dr. Edmund Weitz. All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(in-package :cl-user)

(defpackage :cl-interpol
  (:nicknames :interpol)
  (:use :cl :cl-unicode :cl-ppcre)
  (:export :enable-interpol-syntax
           :disable-interpol-syntax
           :*list-delimiter*
           :*outer-delimiters*
           :*inner-delimiters*
	   :*optional-delimiters-p*
           :*interpolate-format-directives*
	   :interpol-reader))


specials.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL-INTERPOL; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/cl-interpol/specials.lisp,v 1.12 2008/07/23 13:58:40 edi Exp $

;;; Copyright (c) 2003-2008, Dr. Edmund Weitz. All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(in-package :cl-interpol)

(defvar *list-delimiter* #\Space
  "What is inserted between the elements of a list which is
interpolated by #\@.")

(defvar *inner-delimiters* '((#\( . #\))
                             (#\{ . #\})
                             (#\< . #\>)
                             (#\[ . #\]))
  "Legal delimiters for interpolation with #\$ and #\@.")
                             
(defvar *outer-delimiters* '((#\( . #\))
                             (#\{ . #\})
                             (#\< . #\>)
                             (#\[ . #\])
                             #\/ #\| #\" #\' #\#)
  "Legal outer delimiters for CL-INTERPOL strings.")

(defvar *regex-delimiters* '(#\/)
  "Outer delimiters which automatically enable regex mode.")

(defvar *unicode-aliases*
  (make-hash-table :test #'equalp)
  "A hash table which maps Unicode aliases to their real names.")

(defvar *optional-delimiters-p* nil
  "Whether text following $ or @ should interpolate even without a
following delimiter.  Lexical variables are handled correctly,
but the rules are somewhat complex -- see the docs for details.")

(defmacro defvar-unbound (variable-name documentation)
  "Like DEFVAR, but the variable will be unbound rather than getting
an initial value.  This is useful for variables which should have no
global value but might have a dynamically bound value."
  ;; stolen from comp.lang.lisp article <k7727i3s.fsf@comcast.net> by
  ;; "prunesquallor@comcast.net"
  `(eval-when (:load-toplevel :compile-toplevel :execute)
    (defvar ,variable-name)
    (setf (documentation ',variable-name 'variable)
            ,documentation)))

(defvar-unbound *saw-backslash*
  "Whether we have to re-process an \L or \U because it closes several
scopes.")

(defvar-unbound *pair-level*
  "")

(defvar-unbound *stream*
  "Bound to the stream which is read from while parsing a string.")

(defvar-unbound *start-char*
  "Bound to the opening outer delimiter while parsing a string.")

(defvar-unbound *term-char*
  "Bound to the closing outer delimiter while parsing a string.")

(defvar *previous-readtables* nil
  "A stack which holds the previous readtables that have been pushed
here by ENABLE-INTERPOL-SYNTAX.")

(defvar-unbound *readtable-copy*
  "Bound to the current readtable if it has to be temporarily
modified.")

;; stuff for Nikodemus Siivola's HYPERDOC
;; see <http://common-lisp.net/project/hyperdoc/>
;; and <http://www.cliki.net/hyperdoc>

(defvar *hyperdoc-base-uri* "http://weitz.de/cl-interpol/")

(let ((exported-symbols-alist
       (loop for symbol being the external-symbols of :cl-interpol
             collect (cons symbol
                           (concatenate 'string
                                        "#"
                                        (string-downcase symbol))))))
  (defun hyperdoc-lookup (symbol type)
    (declare (ignore type))
    (cdr (assoc symbol
                exported-symbols-alist
                :test #'eq))))
               

load.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL-USER; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/cl-interpol/load.lisp,v 1.1 2003/10/19 15:56:51 edi Exp $

;;; Copyright (c) 2002-2003, Dr. Edmund Weitz.  All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(in-package #:cl-user)

(defparameter *cl-interpol-base-directory*
  (make-pathname :name nil :type nil :version nil
                 :defaults (parse-namestring *load-truename*)))

(loop for file in '("packages"
                    "specials"
                    "util"
                    "unicode"
                    "read")
      do (let ((pathname (make-pathname :name file :type "lisp" :version nil
                                        :defaults *cl-interpol-base-directory*)))
           #-:cormanlisp
           (let ((compiled-pathname (compile-file-pathname pathname)))
             (unless (probe-file compiled-pathname)
               (compile-file pathname))
             (setq pathname compiled-pathname))
           (load pathname)))





